ceno_rt/src/mmio.rs

```diff
@@ -1,110 +1,117 @@
 //! Memory-mapped I/O (MMIO) functions.
 
-use rkyv::{
-    Archived, Deserialize, Portable,
-    api::high::{HighDeserializer, HighValidator},
-    bytecheck::CheckBytes,
-    rancor::Failure,
-};
-
-use core::slice::from_raw_parts;
-
-/// The memory region with our hints.
-///
-/// Logically, this is a static constant, but the type system doesn't see it that way.
-/// (We hope that the optimiser is smart enough to see that it is a constant.)
-fn hints_region<'a>() -> &'a [u8] {
-    unsafe extern "C" {
-        /// The address of this variable is the start of the hints ROM.
-        ///
-        /// It is defined in the linker script.  The value of this variable is undefined.
-        static _hints_start: u8;
-        /// The _address_ of this variable is the length of the hints ROM
-        ///
-        /// It is defined in the linker script.  The value of this variable is undefined.
-        static _hints_length: usize;
-    }
-    unsafe {
-        let hints_length = &_hints_length as *const usize as usize;
-        from_raw_parts(&_hints_start, hints_length)
-    }
+use ceno_serde::from_slice;
+use core::{cell::UnsafeCell, ptr, slice::from_raw_parts};
+use serde::de::DeserializeOwned;
+
+struct RegionState {
+    next_len_at: *const usize,
+    next_data_at: *const u8,
+    alignment: usize,
+    initialized: bool,
 }
 
-/// Get the length of the next hint
-fn hint_len() -> usize {
-    unsafe extern "C" {
-        /// The address of this variable is the start of the slice that holds the length of the hints.
-        ///
-        /// It is defined in the linker script.  The value of this variable is undefined.
-        static _lengths_of_hints_start: usize;
+impl RegionState {
+    const fn new() -> Self {
+        Self {
+            next_len_at: ptr::null(),
+            next_data_at: ptr::null(),
+            alignment: 1,
+            initialized: false,
+        }
     }
-    static mut NEXT_HINT_LEN_AT: *const usize = &raw const _lengths_of_hints_start;
-    unsafe {
-        let len: usize = core::ptr::read(NEXT_HINT_LEN_AT);
-        NEXT_HINT_LEN_AT = NEXT_HINT_LEN_AT.add(1);
+
+    unsafe fn ensure_initialized(&mut self, len_start: *const usize, data_start: *const u8) {
+        if self.initialized {
+            return;
+        }
+        let mut cursor = len_start;
+        let data_offset = unsafe { ptr::read(cursor) };
+        cursor = unsafe { cursor.add(1) };
+        self.alignment = unsafe { ptr::read(cursor) };
+        cursor = unsafe { cursor.add(1) };
+        self.next_len_at = cursor;
+        self.next_data_at = unsafe { data_start.add(data_offset) };
```

Comment:
nitpick: add a link to the function `finalize()` in `ceno_host/src/lib.rs` so that whenever we make any change here, we remember to reflect it there.